### DISCUSSION TITLE
How to contribute

### DISCUSSION THREAD

**Discussion by jpursey** [christofmuc/KnobKraft-orm#249] [General]
How would you prefer minor features, fixes, etc. be done? This project definitely has the core of what I'm looking for, but I'm running into little usability issues, inconsistencies, and bugs. I could either report each of these as issues as they come up or I am more than happy to jump in and make fixes myself (or both). For the latter (performing the fix/change), is the best way to have a bunch of small pull requests from a branch? Or is there a way I can contribute with code reviews more directly? Would you prefer a discussion up front beforehand for minor things, or just discuss in the change itself? I haven't collaborated outside of the very occasional pull request on GitHub, so not sure if there is an established flow I can/should follow....

**Comment by christofmuc** [comment]
@jpursey Happy to hear that! What I'd suggest is to work with pull requests - I made you collaborator on the repo so you don't have to use a fork but can actually push branches right into this repo so you get the Github Actions build pipeline. I'd suggest I do code reviews first because it is an unknown codebase for you.

I think it'd be good to do rather multiple smaller changes/PRs instead of bigger ones, because these are way easier to review and discuss.

With collaborator access, you can also add and edit issues, so if you encounter something you want to track you can not only create issues, but label them/close them when done.

Regarding topics, there is many things going on in parallel, and everytime I try to focus on the next feature or UI improvement, new synth requests come up and the whole adaptation capabilities and testing mechanism gobbles up time :-)

So out of head, the following epics are currently underway:

  * Improving the capabilities of the adaptation interface to support more complex synths like the MKS-70, Until now, those had to be implemented in C++ and not in Python.
  * Improving and documenting testing of adaptations. The advantage with the adaptation model is that they can be nearly completely separate from the KnobKraft main program first, and then only the communication with the synth needs to be tested. This includes a new adaptation interface which is Python class based and not function based, which could make it easier to provide base classes.
  * Finishing the bank management feature released with 2.0 (which is why 2.0 is still labeled beta) - this in general works, but the UI has a lot of quirks and is not easy to understand.
  * Support for individual parameters and external controllers. I have code ready to support BCR2000 and Electra One external controls with full dual sync capability, but never get around to really finish this so it would work for others
  * Any of the other open issues

And of course, anything that interests you! After all, it is a hobby project and should be fun and solve problems you have!

**Comment by jpursey** [comment]
Sounds good, I'll do something quite small at first. First, it involves me trying it from latest rather than the released beta -- maybe you've already fixed some things I've run into. The ones that I'm likely to dig into first is patch name issues as that is what I am primarily running into.

**Comment by christofmuc** [comment]
Yep, start small and let me know when I can help!

**Comment by jpursey** [comment]
Didn't have much time lately, but I did finally get it building after fighting with python installs (independent of this project). The only issue that remained was for whatever reason KnobKraft-orm\builds\_deps\pythonembedded-src contained python38.dll, but the version of python I have installed is the latest (3.11). so the post-build command in The-Orm/CMakeLists.txt failed the copy as it tried to copy python311.dll instead. If I manually change

`"${PYTHON_EMBEDDED_PATH}/python${PYTHON_VERSION_MAJOR}${PYTHON_VERSION_MINOR}.dll"`

with

`"${PYTHON_EMBEDDED_PATH}/python38.dll"`

It all works. I haven't dug any deeper to see what the correct fix should be...

**Comment by jpursey** [comment]
Oh.... it looks like it is hard-coded in the root level CMakeLists.txt. It should probably be hard coded both places or in neither place (the latter being prefereable)?

```
FetchContent_Declare(
  pythonembedded
  URL https://www.python.org/ftp/python/3.8.10/python-3.8.10-embed-amd64.zip
  #URL https://www.python.org/ftp/python/3.10.2/python-3.10.2-embed-amd64.zip
)
FetchContent_MakeAvailable(pythonembedded)
set(PYTHON_EMBEDDED_PATH "${pythonembedded_SOURCE_DIR}")
```

**Reply by jpursey** [reply]
Ok... posted a pull request.
